[WPE][Skia] Implement support for canvas toDataURL

carlosgcampos opened the pull request:
#### 56d0998a1ebb77d9b7e3f6948de2336de044d6a0
<pre>
[WPE][Skia] Implement support for canvas toDataURL
<a href="https://bugs.webkit.org/show_bug.cgi?id=269362">https://bugs.webkit.org/show_bug.cgi?id=269362</a>

Reviewed by Adrian Perez de Castro.

Implement encodeData().

* Source/ThirdParty/skia/CMakeLists.txt: Add JPEG and WebP encoder files.
* Source/WebCore/platform/MIMETypeRegistry.cpp:
(WebCore::MIMETypeRegistry::createMIMETypeRegistryThreadGlobalData): Mark JPEG,
PNG and WebP as formats support for encoding when building with Skia.
* Source/WebCore/platform/graphics/skia/ImageBufferUtilitiesSkia.cpp:
(WebCore::encodeAcceleratedImage):
(WebCore::encodeUnacceleratedImage):
(WebCore::encodeData):
* Source/cmake/FindWebP.cmake: Also look for mux library if requested as component

Canonical link: <a href="https://commits.webkit.org/274723@main">https://commits.webkit.org/274723@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0bfdf6002211e1bf7b7da72518ba87e1716ae6ec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39844 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18855 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42219 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42388 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42151 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21758 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16185 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40418 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15890 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34449 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13750 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35405 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43667 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33297 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36180 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35741 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39511 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39470 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14659 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12057 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16278 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46478 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16326 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9558 "Found 4 new JSC stress test failures: stress/sampling-profiler-microtasks.js.default, stress/sampling-profiler-microtasks.js.dfg-eager, stress/sampling-profiler-microtasks.js.dfg-eager-no-cjit-validate, stress/sampling-profiler-microtasks.js.no-llint (failure)") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5245 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15922 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -86,8 +87,11 @@ if (DEVELOPER_MODE AND (SKIA_ROOT_DIR OR NOT "$ENV{SKIA_ROOT_DIR}" STREQUAL ""))
         OpenGL::GLES
         PNG::PNG
+        WebP::libwebp
+        WebP::mux
```
Isn't it enough to list `WebP::mux`? The imported target should automatically depend on `WebP::libwebp`, which means we would want the following in `FindWebP.cmake`:

```cmake
target_link_libraries(WebP::demux INTERFACE WebP::libwebp)
target_link_libraries(WebP::mux INTERFACE WebP::libwebp)
```

and then here:

```suggestion
        WebP::mux
```

carlosgcampos replied:
I don't know, but I think I prefer to be explicit here, skia doesn't depend only on mux component, but on webp and the mux component. ImageDecoders.cmake also lists both.

aperezdc replied:
Using `INTERFACE` when declaring the dependency (as suggested) causes CMake to include the compiler and linker flags for both when using the `WebP::mux` target. There is no ambiguity, adding `WebP::libwebp` explicitly here doesn't add any value. 

The fact that `ImageDecoders.cmake` uses both is a mistake, and only needed because the dependency declaration is presently missing in `FindWebP.cmake`.

aperezdc commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -860,7 +867,10 @@ target_link_libraries(Skia PRIVATE
+    JPEG::JPEG
     PNG::PNG
+    WebP::libwebp
+    WebP::mux
```
Same here, add only `WebP::mux`.

aperezdc commented on Source/cmake/FindWebP.cmake:
```diff
@@ -154,13 +179,23 @@ if (WebP_DEMUX_LIBRARY AND NOT TARGET WebP::demux)
+        IMPORTED_LOCATION "${WebP_MUX_LIBRARY}"
+        INTERFACE_COMPILE_OPTIONS "${WebP_COMPILE_OPTIONS}"
+        INTERFACE_INCLUDE_DIRECTORIES "${WebP_INCLUDE_DIR}"
+    )
```
As suggested above:

```suggestion
    )
    target_link_libraries(WebP::mux INTERFACE WebP::libwebp)
```

aperezdc commented on Source/ThirdParty/skia/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@ set_property(DIRECTORY . PROPERTY FOLDER "skia")
 # Skia dependencies not used directly in WebKit.
 find_package(Freetype 2.9.0 REQUIRED)
 find_package(Fontconfig 2.13.0 REQUIRED)
+find_package(WebP REQUIRED COMPONENTS demux mux)
```
Here it should be enough to list the `mux` component, because the `demux` one is not used in the rest of the `CMakeLists.txt` file.

```suggestion
find_package(WebP REQUIRED COMPONENTS mux)
```

carlosgcampos replied:
Right, I copy pasted and forgot to remove the demux component.

aperezdc reviewed (changes requested):
Code LGTM, I have left a couple of comments about the CMake bits.

carlosgcampos committed:
[WPE][Skia] Implement support for canvas toDataURL
https://bugs.webkit.org/show_bug.cgi?id=269362

Reviewed by Adrian Perez de Castro.

Implement encodeData().

* Source/ThirdParty/skia/CMakeLists.txt: Add JPEG and WebP encoder files.
* Source/WebCore/platform/MIMETypeRegistry.cpp:
(WebCore::MIMETypeRegistry::createMIMETypeRegistryThreadGlobalData): Mark JPEG,
PNG and WebP as formats support for encoding when building with Skia.
* Source/WebCore/platform/graphics/skia/ImageBufferUtilitiesSkia.cpp:
(WebCore::encodeAcceleratedImage):
(WebCore::encodeUnacceleratedImage):
(WebCore::encodeData):
* Source/cmake/FindWebP.cmake: Also look for mux library if requested as component

Canonical link: https://commits.webkit.org/274723@main
